x-pack/legacy/plugins/siem/server/lib/detection_engine/routes/utils.ts

```diff
@@ -215,13 +221,61 @@ export const transformBulkError = (
   }
 };
 
-export const getIndex = (getSpaceId: () => string, config: LegacyServices['config']): string => {
-  const signalsIndex = config().get<string>(`xpack.${APP_ID}.${SIGNALS_INDEX_KEY}`);
-  const spaceId = getSpaceId();
+export const buildRouteValidation = <T = {}>(schema: Joi.Schema): RouteValidationFunction<T> => (
+  payload: T,
+  { ok, badRequest }
+) => {
+  const { value, error } = schema.validate(payload);
+  if (error) {
+    return badRequest(error.message);
+  }
+  return ok(value as T); // TODO: infer type from our schema
+};
 
-  return `${signalsIndex}-${spaceId}`;
+const statusToErrorMessage = (statusCode: number) => {
+  switch (statusCode) {
+    case 400:
+      return 'Bad Request';
+    case 401:
+      return 'Unauthorized';
+    case 403:
+      return 'Forbidden';
+    case 404:
+      return 'Not Found';
+    case 409:
+      return 'Conflict';
+    case 500:
+      return 'Internal Error';
+    default:
+      return '(unknown error)';
+  }
 };
 
+export class SiemResponseFactory {
+  constructor(private response: KibanaResponseFactory) {}
+
+  error<T>({ statusCode, body, headers }: CustomHttpResponseOptions<T>) {
+    const defaultedHeaders = {
+      'content-type': 'application/json',
+      ...headers,
+    } as CustomHttpResponseOptions<T>['headers']; // https://github.com/microsoft/TypeScript/issues/27273
```

Comment:
You can do this if you don't want this `as cast` here:

```ts
  error<T>({ statusCode, body, headers }: CustomHttpResponseOptions<T>) {
    const contentType: CustomHttpResponseOptions<T>['headers'] = {
      'content-type': 'application/json',
    };
    const defaultedHeaders: CustomHttpResponseOptions<T>['headers'] = {
      ...contentType,
      ...(headers ?? {}),
    };
```

It makes sense that TypeScript is trying to be extra careful in some parts but requires more ceremony in order to get your types exactly correct without resorting to a cast. Bummer that it requires an additional check on something that is undefined for that spread but, at least the `??` helps out in that case.

Reply:
230ae006e91a5c41db891fae7ffb1b2907b13dd4